File: Cpp/strategy.cpp
#include <iostream>
#include <vector>
#include <algorithm>

class SortStrategy {
public:
    virtual ~SortStrategy() {}
    virtual void sort(std::vector<int>& data) = 0;  // Pure virtual function
};

class BubbleSortStrategy : public SortStrategy {
public:
    void sort(std::vector<int>& data) override {
        int i, j;
        for (i = 0; i < data.size()-1; i++)
            for (j = 0; j < data.size()-i-1; j++)
                if (data[j] > data[j+1])
                    std::swap(data[j], data[j+1]);
    }
};

class QuickSortStrategy : public SortStrategy {
public:
    void sort(std::vector<int>& data) override {
        quickSort(data, 0, data.size() - 1);
    }

private:
    int partition(std::vector<int>& data, int low, int high) {
        int pivot = data[high];
        int i = (low - 1);

        for (int j = low; j <= high - 1; j++) {
            if (data[j] < pivot) {
                i++;
                std::swap(data[i], data[j]);
            }
        }
        std::swap(data[i + 1], data[high]);
        return (i + 1);
    }

    void quickSort(std::vector<int>& data, int low, int high) {
        if (low < high) {
            int pi = partition(data, low, high);
            quickSort(data, low, pi - 1);
            quickSort(data, pi + 1, high);
        }
    }
};

class SortingSystem {
public:
    SortingSystem(SortStrategy* strategy) : strategy(strategy) {}
    ~SortingSystem() { delete strategy; }

    void sort(std::vector<int>& data) {
        strategy->sort(data);
    }

private:
    SortStrategy* strategy;
};

int main() {
    std::vector<int> data = {64, 34, 25, 12, 22, 11, 90};

    SortingSystem* system = new SortingSystem(new BubbleSortStrategy());
    system->sort(data);
    for(int i : data)
        std::cout << i << " ";
    delete system;

    std::cout << std::endl;

    data = {64, 34, 25, 12, 22, 11, 90};  // Reset the data
    system = new SortingSystem(new QuickSortStrategy());
    system->sort(data);
    for(int i : data)
        std::cout << i << " ";
    delete system;

    return 0;
}

File: Python/strategy.py
from abc import ABC, abstractmethod
from typing import List


class SortStrategy(ABC):
    @abstractmethod
    def sort(self, data: List[int]) -> List[int]:
        pass


class BubbleSortStrategy(SortStrategy):
    def sort(self, data: List[int]) -> List[int]:
        data = data.copy()
        n = len(data)
        for i in range(n):
            for j in range(0, n-i-1):
                if data[j] > data[j+1] :
                    data[j], data[j+1] = data[j+1], data[j]
        return data


class QuickSortStrategy(SortStrategy):
    def sort(self, data: List[int]) -> List[int]:
        data = data.copy()
        self.quick_sort(data, 0, len(data) - 1)
        return data

    def partition(self, data: List[int], low: int, high: int) -> int:
        i = low - 1
        pivot = data[high]
        for j in range(low, high):
            if data[j] <= pivot:
                i += 1
                data[i], data[j] = data[j], data[i]
        data[i+1], data[high] = data[high], data[i+1]
        return i+1

    def quick_sort(self, data: List[int], low: int, high: int):
        if low < high:
            pi = self.partition(data, low, high)
            self.quick_sort(data, low, pi - 1)
            self.quick_sort(data, pi + 1, high)


class SortingSystem:
    def __init__(self, strategy: SortStrategy):
        self.strategy = strategy

    def sort(self, data: List[int]) -> List[int]:
        return self.strategy.sort(data)


if __name__ == "__main__":
    # Użycie
    bubble_sort_strategy = BubbleSortStrategy()
    quick_sort_strategy = QuickSortStrategy()

    data = [64, 34, 25, 12, 22, 11, 90]

    system = SortingSystem(bubble_sort_strategy)
    print(system.sort(data))  # Sortuje danymi używając strategii Bubble Sort

    system = SortingSystem(quick_sort_strategy)
    print(system.sort(data))  # Sortuje danymi używając strategii Quick Sort
